node/src/components/chainspec_loader.rs

```diff
@@ -249,32 +300,108 @@ impl ChainspecLoader {
             root_dir,
             reactor_exit,
             next_upgrade,
+            maybe_immediate_switch_block_data: None,
         };
 
         (chainspec_loader, effects)
     }
 
     fn handle_initialize<REv>(
         &mut self,
-        _effect_builder: EffectBuilder<REv>,
-        maybe_highest_block: Option<Box<BlockHeader>>,
+        effect_builder: EffectBuilder<REv>,
+        maybe_highest_block_header: Option<Box<BlockHeader>>,
     ) -> Effects<Event>
     where
-        REv: Send,
+        REv: From<Event> + From<ContractRuntimeRequest> + Send,
     {
-        self.reactor_exit = Some(
-            Self::should_exit_for_upgrade(
-                maybe_highest_block,
-                self.next_upgrade_activation_point(),
-            )
-            .then(|| ReactorExit::ProcessShouldExit(ExitCode::Success))
-            .unwrap_or(ReactorExit::ProcessShouldContinue),
+        // Check if we're not running a version that's already outdated - if it is, we should exit
+        // and upgrade.
+        if Self::should_exit_for_upgrade(
+            maybe_highest_block_header.as_deref(),
+            self.next_upgrade_activation_point(),
+        ) {
+            self.reactor_exit = Some(ReactorExit::ProcessShouldExit(ExitCode::Success));
+            return Effects::new();
+        }
+
+        let current_chainspec_activation_point =
+            self.chainspec.protocol_config.activation_point.era_id();
+
+        match maybe_highest_block_header {
+            Some(header)
+                if header.is_switch_block()
+                    && header.era_id().successor() == current_chainspec_activation_point =>
+            {
+                // This is a valid run immediately after upgrading the node version, we'll need to
+                // create an immediate switch block.
+                trace!("valid run immediately after upgrade");
+                let upgrade_config_result =
+                    self.new_upgrade_config(&header, Arc::clone(&self.chainspec_raw_bytes));
+                async move {
+                    match upgrade_config_result {
+                        Ok(upgrade_config) => (
+                            header,
+                            effect_builder
+                                .upgrade_contract_runtime(upgrade_config)
+                                .await,
+                        ),
+                        Err(error) => (header, Err(error.into())),
```

Comment:
Is it necessary to pass `header` through this `async` block? Could it just be part of the closure below, that is passed to `event`?

Reply:
Looks like it could - I'll try it :+1: